[3.x] Allow exporting custom resources from/to any scripting language (GDScript, VisualScript, C#, NativeScript, PluginScript)

willnationsdev opened the pull request:
This PR adds the ability to export custom resources and arrays of custom resources of any language's type *from* an exported property defined in any language.

Closes godotengine/godot-proposals#18, at least, in Godot 3.2.

~~This builds on top of the commit in #44796, so that would have to be merged first.~~ This now includes the PR content of #44796 as of the 3.x rebase. For a detailed breakdown of the features commit-by-commit, see the 4.0 version at #48201.

This is a continuation of work started by me and @JFonS and would replace his 3.2 PR which deals exclusively with GDScript (#41264).

It includes editor changes to support script classes as valid resource exports.

Because the definition of PropertyInfo objects is manual in VisualScript, NativeScript, and PluginScript, the feature works out-of-the-box for those languages.

GDScript support was already implemented by JFonS's PR, but I've incorporated those changes and more.

C# support is added as well. The `[Export]` attribute now supports the following forms...

```cs
// Infer hint and hint string (no changes)
[Export]

// Manually define hint and hint string. Only applies if valid hint cannot be calculated (no changes).
[Export(PropertyHint.None, "")]

// Applies custom resource name if applicable to the member's data type (new!).
// Only necessary when exporting a Resource or List<Resource> for a non-C# resource type.
[Export("ClassName")]
```

Demo Project:
[Demo_GP22_3_2_Resources.zip](https://github.com/godotengine/godot/files/5760748/Demo_GP22_3_2_Resources.zip)

Screenshots:

GDScript usage, exported properties contracted:
![contracted_resources_gd_usage](https://user-images.githubusercontent.com/16217563/103464545-2159cb80-4cfa-11eb-8c22-e126ab81df31.png)

GDScript usage, exported properties expanded:
![expanded_resources_gd_usage](https://user-images.githubusercontent.com/16217563/103464551-2c146080-4cfa-11eb-8f69-8938eabfc8f7.png)

Note I also have `export var gd_data: GDData` syntax supported as well (so you can infer the export type from the GDScript type hint).

C# usage doing the same thing:
![csharp_export_resources_gdlist](https://user-images.githubusercontent.com/16217563/103464559-3a627c80-4cfa-11eb-96ee-06cda26c2549.png)

willnationsdev committed:
Add ScriptServer and export all custom resources.

vitorbalbio commented:
This qualifies Godot as a tier one engine for data intensive games. Congrats @willnationsdev, @JFonS and all devs involved.

willnationsdev commented:
@vitorbalbio I will say that, for simulation-heavy games that *also* require a lot of data, this alone won't really be enough. This is merely a big usability improvement for general-purpose games that won't data-driven behavior. To get a significant runtime performance improvement (i.e. processing over large amounts of data), you'd have to resort to writing module code with a custom CPU cache-efficient code with your own Server implementations, etc.

@AndreaCatania is doing great work in this domain as he is writing a module that introduces an ECS framework that can run alongside the Godot Engine runtime and cooperates with the scripting API. He has seen [major performance improvements](https://twitter.com/_AndreaCatania/status/1312824560358752256?s=20) for any simulation logic he moves over to his module's systems, so if you need a game like that, keep an eye on his development. He has [multiple customizable system pipelines](https://twitter.com/_AndreaCatania/status/1342063206597013506) defined with [editor tools](https://twitter.com/_AndreaCatania/status/1340250917317259267?s=20). Time will tell if the developers ever decide to actually integrate his ECS module with the core engine (*they absolutely should though*).

vitorbalbio commented:
> @vitorbalbio I will say that, for simulation-heavy games that _also_ require a lot of data, this alone won't really be enough. This is merely a big usability improvement for general-purpose games that want data-driven behavior. To get a significant runtime performance improvement (i.e. processing over large amounts of data), you'd have to resort to writing module code with custom CPU cache-efficient code with your own Server implementations, etc.
> 
> @AndreaCatania is doing great work in this domain as he is writing a module that introduces an ECS framework that can run alongside the Godot Engine runtime and cooperates with the scripting API. He has seen [major performance improvements](https://twitter.com/_AndreaCatania/status/1312824560358752256?s=20) for any simulation logic he moves over to his module's systems, so if you need a game like that, keep an eye on his development. He has [multiple customizable system pipelines](https://twitter.com/_AndreaCatania/status/1342063206597013506) defined with [editor tools and script code](https://twitter.com/_AndreaCatania/status/1340250917317259267?s=20). Time will tell if the developers ever decide to actually integrate his ECS module with the core engine (_they absolutely should though_).

I mean more like RPGs and Tactical games that you need to define lots of data/status for large amount of items, objects and so on. The same use cases where Scripted Objects of Unity shines.

willnationsdev commented:
@vitorbalbio Ah, yeah. I really would like to see a proper spreadsheet class added to the engine/Editor to simplify working with tabular data with full static typing and all the custom editors from the Editor integrated out-of-the-box. godotengine/godot-proposals#13 That would make it much easier to just create large amounts of data for projects like those.

willnationsdev commented:
Fixed a bug where type hints for GDScript inner classes were mistakenly not parsing to the end of the line anymore.

willnationsdev commented:
Note that someone on Reddit reported a bug with this where referencing an inner class within the same file using a relative symbol name doesn't work (even though I fixed it so the fully-qualified name in other scripts works).

willnationsdev commented:
Fixed a bug where script classes that were part of a nested `res://addons/*` directory weren't showing up in the CreateDialog unless they were part of an enabled plugin. I've now changed the CreateDialog code to explicitly check for the presence of a `plugin.cfg` file within the directory, and if there isn't one, it will continue to display the script class in the tree of types.

willnationsdev commented:
Fixed a bug where, once you had assigned a script class to a node, there was no way to remove the script class from the node because the `Change Type` operation would preserve the `script` property value even if the node you were "changing type" to was explicitly the instance base type.

willnationsdev commented:
Added a warning message when explicitly removing a built-in script. Built-in scripts' inheritance cannot be determined because they are never compiled, ergo their `get_base_script()` method won't produce a result (at least, in GDScript). Added a `WARN_PRINT_ONCE` so that users are not alarmed/confused as to why the entire script was removed rather than removing just down to the script class like it should (mimicking the custom type feature).

DmitriySalnikov commented:
what is the development status of this PR?

fire commented:
I'm not working in 3.2 so I'll let some else review.

willnationsdev commented:
@DmitriySalnikov It's status is that no one has thus far reported any issues that I have not fixed, so it's probably in a beta state or something. We need more people to use it / review it to confirm if there are any unforeseen problems.

davcri commented:
Some days ago I tried https://github.com/willnationsdev/godot/commit/9ff013e7c0d66ea8d6a0a659ea96a866b08e78a2 with a [small game](https://github.com/crystal-bit/defending-todot) that used custom resources for enemies and turrets. Everything was working as expected. Unfortunately I didn't push the branch that I used to test the new features.

https://github.com/godotengine/godot/issues/41442 was still an issue for me, but I don't know if it's related to this PR.

Arrow-x commented:
So I have been working with this for a bit and love it, is it going to be merged for the next RC?

akien-mga commented:
No, but we can merge it for the first 3.2.5 beta if I can get some contributors to review it :)

xDGameStudios commented:
would really like to see this merged.. is great news!

Zaya6 commented:
Working on a game that could really use this, hopefully it can get reviewed soon

rosshadden commented:
I'm really glad I found this because I was already spending time trying to implement a different fix for the same problem. Your solution is much more generic and elegant. Finding it now prevented me from spending a _lot_ more time on my approach :sweat_smile:.

@willnationsdev Feedback:
This works great! I haven't found any issues with it so far however I do have a question (which may actually be an issue or request in disguise). Is there a way to set a default value for these custom resources?

I am working with my own simple experiment using your branch to test the boundaries, and to me it seemed natural to try to set a default value. I tried two main things.

Note that the static typing in these examples is more verbose/redundant than it needs to be, and is not really part of the problem.
But I only get one of the lint errors I am going to mention if I annotate the property as `: GdData`.

## Exporting an instance

This makes the most sense to me logically because you are setting the variable to a new instance of the type, which is what I believe is expected.
However this shows as `[empty]` in the inspector and is not editable.

```gdscript
export(GdData) var gdData: GdData = preload("res://GdData.gd").new()
```

## Exporting the class itself

This makes less sense, but the linting error it gives is interesting:

> The assigned expression's type (GdData) doesn't match the variable's type (GdData)

I suspect despite being stringified the same, the former is actually "GdData class" and the latter is "GdData instance" but you would know better.

```gdscript
export(GdData) var gdData: GdData = preload("res://GdData.gd")
```

## Closing

I guess it kind of doesn't make sense to statically set a default export variable to a new instance because that might imply the instance should be shared across all of them... but I can't really think of any other ways to set a default and was wondering if you could suggest anything.

Thank you very much for your time on this, it will make working on games the way I am used to _much_ more sane! <3

willnationsdev commented:
@rosshadden 

> This works great!

That's good to hear. In particular though, I haven't heard much feedback from folks about the way the script classes now work in the SceneTreeDock. Could you experiment with adding scripts, removing scripts, changing types, and doing all the same with anonymous scripts that derive script classes too? Get a feel for how the new relationships work and tell me what you think?

The first commit, associated with a different PR, changes the way script classes work so that the SceneTreeDock treats them more like built-in C++ classes, and no one has really commented on that so far, despite probably being the most controversial aspect of the changes (and which are totally subject to change).

> However this shows as [empty] in the inspector and is not editable.

I would've expected your "exporting an instance" example to work. I would've also expected this to work:

```gdscript
export var gdData := GdData.new()
```

If not though, then that means it's a bug and needs to be fixed. That other bug you mention sounds like it could be the culprit though.

> I suspect despite being stringified the same, the former is actually "GdData class" and the latter is "GdData instance" but you would know better.

This is correct. We could probably make the error message in that case a little bit more clear.

> Finding it now prevented me from spending a lot more time on my approach
> 
> Thank you very much for your time on this

Yeah, I spent a *lot* of time on this approach myself (these two commits are, like, a year's collective research/development/debugging in my off time, lol). Glad it's finally helping people.

willnationsdev commented:
@rosshadden Another thing you could try (although this might be not really what you want) is to define a setter for the variable so that whenever someone attempts to assign it to `null`, it instead gets assigned a new instance of the class. That will then make the default initialization always populate it with a ready-to-edit object in the Inspector. Though...I haven't actually tried that and don't know if that could cause any problems.

rosshadden commented:
@willnationsdev I played with your changes a lot more, in particular some of the things you mentioned like using script classes in the SceneTreeDock. For example I created a "New Resource" and chose a custom class as the resource type. After doing so all normal operations with the resulting `.tres` file work as I expected. The `type` stored in the tres file is "Resource" and not the specific class I chose, but this isn't untrue as they all extend Resource anyway.

I actually haven't encountered any issues with this so far other than the one I mentioned above where you cannot set a default (again, seems to be a problem in `3.2` in general, not introduced by your changes). I did try your other suggestions for setting defaults as well as a few more and none of them worked. Nor did creating a setter.

I also experimented with nesting resources which worked as well. I haven't tried nesting a resource as a property inside itself, but that's rather silly. I feel like at this point if there are problems they won't be caught with the workflows I am using this for, and thus I give it my :+1:. Hopefully this can gain traction and others can review as well so we can catch any problems with it and fix them.

BTW the approach I mentioned I was working on before I found this PR was adding a "Make instances unique" menu option at the resource level which would basically do what clicking "Make Unique" does per resource instance (the `OBJ_MENU_MAKE_UNIQUE` resource menu option). The benefit being you could set this in one place, on a resource, instead of having to click it every time you use a certain resource. In my experience most of the time you want a resource to be unique you want it to be unique everywhere.
This PR does essentially exactly this in effect, because the end result of both approaches is that the properties of the resource are stored in the scene file in question.

red1939 commented:
Basing on @akien-mga comment it seems that this pull request is frozen until some good soul (i.e. contributor well-versed in the way of Godot) reviews it (and yes, there is *a bit of* code to be reviewed ;)). That said, I am guessing in order to push it a little closer to RC wagon some testing would be helpful I guess?

akien-mga commented:
It's not going to be merged for 3.2.4 RCs, but yes testing would be very useful. This will likely be reviewed and merged for 3.2.5 betas, but any issues found prior to that and fixed in the PR would be quite useful.

tomasjm commented:
Thank you for this feature!
I was testing it with a project and I think I found a bug.
Specifically with C# when using a [Tool] script trying to access to a property of a custom resource, running the script in editor causes a godot crash.

Here is the [repo](https://github.com/tomasjm/godot_procedural3d_test) of the project I tested.
**I don't tested this project with GDScript, but I tested this behaviour (or the same case) with GDScript and it works well**

## Steps to reproduce:
1. Build project solution.
2. Open ErrorTest.tscn in root folder.
3. Test it by changing some script variables of the parent in the editor and clicking "update" checkbox (**The custom resource is called HeightTerrain**).
3. Open MapUtils.cs in Scripts folder on script editor.
4. [Uncomment line 25](https://imgur.com/a/7Jjt9QN) then build project solution.
5. Run the scene and see the console print -> should work
6. Click "update" checkbox -> should crash

Apparently the crash occurs because the class object is not defined in the editor before running the scene, if we change the color property of the custom class HeightTerrain and change the print of line 25 with heightTerrain.Color it will print a random value like (0,0,0,13e-20). When running the scene it works good.

I don't know if this is my problem so I share it here, I anticipate to apologize if the error was something that I missed.

akien-mga commented:
Would be great to get this merged soon for 3.4, and it should likely also be forward-ported to the `master` branch. Most core refactoring to script languages has happened in `master` already so if this issue hasn't been solved in the process, it still needs to be.

Would be great to have in-depth review by @vnen and @neikeq, and @godotengine/gdnative stakeholders.

Would also be good to figure out the crash reported in https://github.com/godotengine/godot/pull/44879#issuecomment-803851774 before we merge.

aaronfranke commented:
@willnationsdev Do you expect to have a version of this PR available for the master branch? It would make sense to make this change there first, even if it's not implemented for GDNative there yet. It might also be easier to merge since it would be changing fewer things at once.

willnationsdev commented:
Yeah, I could try and make time for that. I wanted it to be in master in the first place, hehe.

willnationsdev commented:
While the branch still says gdres-all-3.2, it has been rebased onto the `3.x` branch. The original 3.2 build, for those who need it, is now in `gdres-all-3.2-backup`.

willnationsdev commented:
@akien-mga Since this and the 4.0 port #48201 are so large, would it be preferred if I make them both draft PRs and then open separate smaller PRs that reference the draft ones for grouping purposes? I'd basically end with one for each commit in the 4.0 port, i.e. 9 PRs.

cgbeutler commented:
@willnationsdev
Curious if this would fix the bug I am seeing, or if I should file a new bug.
If I create a custom resource _from_ C#, the script on it seems to lack a proper `resource_path`, meaning it gets saved to disk wrong, can't load later, and overall has issues. I assume this is because the script resource was not loaded through the ResourceLoader's cache.
In other words:
```csharp
var cust = new CustomResource();
GD.Print($"script path = { ((Script)cust.GetScript())?.ResourcePath }");
```
prints: `script path = `

willnationsdev commented:
Possibly. I'm actually not really sure. Does GDScript print something more reasonable in its equivalent? If so, it could be related. I don't know too much about the inner workings of resource loading and don't have time to investigate atm. Perhaps someone else can chime in?

cgbeutler commented:
Some explanation on Resources loading/saving just in case it relates to this PR:
`ResourceLoader` tries to cache the resources it loads, including scripts. This is basically done by using `resource_path` as a kind of "cache key." If you call `load()` with the same path twice, you get one shared reference.
_Usually_ scripts are loaded from the FileSystem, and are subsequently caching using their file path as their `resource_path`. Scene and resource files use this caching as well when loading an `ext_resource`, like a script.

When creating a custom resource in C# using its constructor, the script appears to be set as a non-cached version. That means when it comes time to save, the custom resource's script gets saved as a `sub_resource` instead of an `ext_resource`.
What gets saved:
```
[sub_resource type="CSharpScript" id=1]
```
What should have been saved:
```
[ext_resource path="res://CustomResource.cs" type="Script" id=2]
```

A subresource in a scene or resource file should contain below it everything needed to create it. An external resource in a scene or resource file simply loads or gets the cached file for the given path.
As you can see above, the `sub_resource` version contains __nothing__. The script cannot be loaded, and even if it could be, it would not be the cached, shared version that the rest of the system would be using. Manually edit the scene file to replace the `sub_resource` with an `ext_resource` and your custom resource will magically start working again!

willnationsdev commented:
That certainly sounds like a solution the problem then if the custom resource works after fixing that issue. The next step would just be figuring out *why* it saves it as a sub resource instead of as an external resource. I previously didn't know the function of `resource_path`, but I knew the resources were cached, so that makes a lot of sense now.

If you have time to figure out the details further, by all means. I'm neck-deep in work stuff atm. XD

cgbeutler commented:
Hmm... looks like it's happening with all objects that have a custom C# script. All those scripts are technically detached from the cached one. I'm going to file it as a separate bug.

phelioz commented:
Have used this PR casually in my project for awhile now, nothing directly to add except that for my project it's working great! (non C# project).
Really awesome work, makes using custom resources so much nicer!

willnationsdev commented:
@phelioz, thanks for the confirmation that it is working well for you in a non-C# project.

willnationsdev commented:
I just pushed a change that includes the XML documentation for all changes to the scripting API introduced by this PR. I also amended the `ScriptServer.get_global_class_map()` implementation so that Dictionary for each class contains not only keys for `"language"`, `"path"`, and `"base"`, but also `"native"` and `"icon_path"`. This way, if you need to avoid making repeated queries to the ScriptServer's methods across the scripting API, you can just grab the entire Dictionary of data and use it locally.

willnationsdev commented:
@cgbeutler Lol, I finally got around to testing out your C# bug. Figured it out. It is indeed a user error. All you had to do was add the `[Tool]` attribute to your `HeightTerrain` C# script, in *addition* to the `[Global]` attribute (`[Global("HeightTerrain")]` is unnecessary). Now that your custom type is also capable of being loaded at editor time, since it's a tool script, your logic now ends up working fine.

willnationsdev commented:
I backported some of the cleanliness updates that I made during the 4.0 port and improved the degree to which the 3.x and 4.0 versions match 1-to-1 (to make comparing the PRs easier). I also identified and fixed a bug where the EditorPropertyResource's menu options wouldn't handle file extensions in cases where multiple classes are provided and one of those classes is a script class.

With this, and identifying the C# issue was a user problem, this PR should be pretty ready to merge whenever it can be properly reviewed.

cgbeutler commented:
@willnationsdev I think you're mixing me up with something else. I have no HeightTerrain class... Also, I didn't forget tool.

Regardless, my thing is a separate bug.

willnationsdev commented:
@cgbeutler Ah, you're right, my bad. It was @tomasjm who reported the C# crash bug. (For context, tomasjm, you just needed to add `[Tool]` to your `HeightTerrain` script)

tomasjm commented:
> @cgbeutler Ah, you're right, my bad. It was @tomasjm who reported the C# crash bug. (For context, tomasjm, you just needed to add `[Tool]` to your `HeightTerrain` script)

omg, thank you, my bad! I had not reviewed the code with the freshest mind to find that haha, hope this PR gets merged, nice work!

mphe commented:
I've been using this patch almost since it was released and haven't had any problems so far.
However, recently I updated to the latest version and loading custom resources in the editor is now broken. Exporting custom resources works fine, it is displayed correctly in the inspector, I can click the "New CustomResource" button and it correctly creates a new resource, but the load function fails with a warning *"The selected resource (Resource) does not match any type expected for this property (GDData)."*
At the moment when I press "Load" and the file-open-dialog shows, the console spams about 30 times *"Cannot get class 'GDData'."*.

Tested with the most recent gdres-all-3.2 branch and the demo project.

cgbeutler commented:
@mphe what gets saved to the tscn file for the ones that won't load? For me, In the past, it has specifically been the script resource that isn't linked properly. (See #48828)

mphe commented:
@cgbeutler Nothing gets saved. The property remains unchanged and keeps the value it had before.

willnationsdev commented:
@mphe 

1. Have you been regularly updating to the latest version of the `gdres-all-3.2` branch thinking that it only applies to the `3.2` tag, cause it actually has been attempting to keep track with latest versions of the 3.x branch of the Godot repo. There is a preserved 3.2 branch with `gdres-all-3.2-backup` if that is what you need.
2. Just to be sure, are you somehow using the sample project I originally published with this that had `GDData` as a resource type?  Or does your project have independent stuff so that an error message like that makes no sense? If the latter, then maybe I somehow over-committed a change that kept info about a demo project's resource type? Don't see how that could have happened though.
3. Which language is your custom resource written in and in which language are you exporting a reference to that resource type? For example, is it a C# class that owns a GDScript resource? Are there certain combinations that do not create problems?

mphe commented:
1. I'm following the 3.x branch and update every 1-3 months or after official releases. Hence, I can't say exactly when this bug appeared, but I think the last time it definitely worked was at the 3.3.2 release. Just for clarity, when updating I create a clean branch from upstream 3.x and merge this `gdres-all-3.2` branch (among others).
2. After I noticed this problem in my project, I cloned and built your `gdres-all-3.2` branch directly and tested it with the demo project from your top post. Since the problem persists, it is unlikely to be related to my project and godot version.
3. I'm only using GDScript. Not tested with C#.

willnationsdev commented:
@mphe Gotcha. Thanks. I'll look into it tonight.

willnationsdev commented:
@mphe 

So, I investigated as much as I could, but I couldn't seem to replicate *any* kind of error similar to the one you described.

Based on the error message, it sounds like there is a property typed for `GDData` that is being assigned a basic `Resource` (or at least, that's what the GDScriptParser appears to think is happening?), so that would indicate to me that there is some `GDData` property info configured somewhere, but the Editor doesn't know what that is, so the assigned GDData instance is interpreted as just a basic Resource, hence the error - or something like that.

However, I did a fresh update from upstream/3.x, rebased the `gdres-all-3.2` branch, and did not encounter any issues with the demo projects' various scenes. I even tried creating new resources, saving them to files, and loading them from files (since you said you clicked 'load', opened the dialog, and got spammed with errors). I saw no errors of that sort when I was messing around.

Not really sure what your issue is since I can't seem to reproduce it.

I *did* notice there was some sort of bug with GDScript's parsing of the `class_name` statement when an existing global GDScript is renamed or moved, even via the Filesystem Dock, but restarting the editor appears to fix that issue (probably just a cache issue in the ScriptServer code).

I also noticed that the newfound EditorResourcePicker breaks the display of the proper class name for the currently selected resource in its dropdown picker, so I made a fix for that and re-pushed. The fix is also present on the 4.0 version of this PR.

mphe commented:
The problem also persists with the new version.

Maybe it is somehow related to a regular, non-Mono build of Godot? Unfortunately, I'm failing to get the Mono version compiling, so I can't test it right now.

willnationsdev commented:
@mphe Alrighty, take 2, and this time I seem to have reproduced the issue. Not sure why I wasn't triggering the errors before, but now they are happening. It's strange though. If you create a built-in, inline instance in the Inspector it works fine. And if you create a standalone custom resource in the Inspector and save it, it's fine. Now, clicking the `Load` button at the *top* of the Inspector for a custom resource works fine (that's what I tested earlier), but clicking the `Load` option in the dropdown beside a specific resource does *not* work, nor does dragging and dropping a resource.

These errors are likely because of the migration to the EditorResourcePicker which had other associated bugs I had to fix and this is just more of the same. I will look into it and get a fix uploaded ASAP.

willnationsdev commented:
@mphe There were indeed some errors related to the EditorResourcePicker that have now been fixed, but there were also two *other* errors in very different places for which there are now meaningful changes/fixes.

1. The EditorData class has a cached lookup table of path-to-name mappings for script classes so that, while you are in the Editor, the engine can look up the name of a global script class based on its filepath. Part of the problem is that this map is only updated *accurately* if the typename also has a corresponding icon. It would be easy enough to correct it so that it works for other global scripts too, however, this is information that could *also* be useful to runtime instances. For example, any apps/games/editors/modding tools that iterate over resources, get the scripts of said resources, and check if the script is a global script class with a typename. Thus, I have moved the reverse lookup map to the ScriptServer and replaced all references to the EditorData one (of which there were only a few) for the ScriptServer one.
2. Because all core ResourceFormatLoaders evaluate every type, even types that are global script classes, I made a slight change to the ResourceLoader so that only ResourceFormatLoaders *WITH* a script evaluate the loaded resource type as given. Otherwise, it converts the stated type to the base C++ type and passes it along. This way, loaders defined in the C++ code only have to worry about C++ resource types and it is only the user-defined loaders that need to account for scripted types.

I will make an update to the 4.0 version with similar changes soon.

mphe commented:
@willnationsdev Great! Everything seems to work again. Thanks for fixing!

>  but clicking the Load option in the dropdown beside a specific resource does not work, nor does dragging and dropping a resource.

That's what I meant. Didn't think of the other load button.

ATHellboy commented:
Is below code possible with this PR ?

```
class Order extends Resource:
	func _init():
		pass

class Wait extends Order:
	export(float) var seconds

class Wander extends Order:
	export(float) var seconds = 0.0

class Towards extends Order:
	export(NodePath) var object

class Interact extends Order:
	export(Resource) var behavior
	
export(Array, Order) var orders
```

Just showing classes which are inherited from Order class.

rosshadden commented:
@willnationsdev Back in February [I brought up](https://github.com/godotengine/godot/pull/44879#issuecomment-787531732) that defaults were not being respected in the inspector and you said "seems to be a problem in 3.2 in general, not introduced by your changes". I just discovered that this seems to have indeed fixed upstream and that bug is no longer present.
When I set a default in code, the default is present as expected in the inspector!

```gdscript
export(Whatever) var whatever = preload("res://path/to/whatever.tres")
```

willnationsdev commented:
FYI, am trying to rebase this, but I found that, for some reason, the Make Unique operation is not working successfully (don't think it ever was(?)), so now I'm trying to fix that before I push an update.

Mastermori commented:
Any idea when this will be merged? This is a life changer!

Calinou commented:
> Any idea when this will be merged? This is a life changer!

The `3.x` version of this pull request will be merged after 3.4 is released to avoid introducing regressions in 3.4.

cgbeutler commented:
If you get a spare moment, could you test if this branch also fixes https://github.com/godotengine/godot/issues/52729 ?  I'm guessing it's a separate issue, but just in case.

willnationsdev commented:
@cgbeutler Sure, I'll give it a shot when I have free time. Kinda busy, but I'll try to find some time.

cgbeutler commented:
@willnationsdev it's looking like an unrelated issue. Ignore my previous request.

LightningAA commented:
3.4 has been released now, it seems like all that's missing before this PR can be merged is fixing the doctool checks and maybe a rebase

Hakunamawatta commented:
What's stopping this from getting merged?

pycbouh commented:
> What's stopping this from getting merged?

Lack of reviews. None of the maintainers for the area have been able to test and code review it yet.

dploeger commented:
Additionally, @willnationsdev needs to implement my changes but I guess he couldn't yet find the time.

Ternvein commented:
I think it's totally needed for 3.5. It's been ages since first pull request with fixes for this issue, and at least 3 such requests was archived because authors tired of updating it for months. If this will happen here too, it will be most regrettable. And this fix really needed.

pycbouh commented:
> It's been ages since first pull request with fixes for this issue, and at least 3 such requests was archived because authors tired of updating it for months.

It's the nature of doing something in the core systems, especially something big. This is a huge PR with a lot of consequences beyond just the ability to pick a custom resource in the inspector. It needs thorough review from the core team, and the core team has been pretty busy these last couple of years.

dploeger commented:
@pycbouh sorry for mentioning you directly, but you recently commented and maybe have a quick comment on this:

In #48201 @willnationsdev and I are talking about me adopting the two PRs. I was think about getting one of the two ready and merged in a new PR and based on the comments there (back-)port to the other version. Would that make sense to you?

pycbouh commented:
@dploeger Feel free to open a new PR based on Will's work. You could also ask him for a write permission on his fork and that way you could push directly to the same branch on his repo and update the PR this way. If not, opening a new one is fine, we'll close one as superseded.

You can keep Will's credit by using the co-authored-by tag in the commit messages.

dploeger commented:
@pycbouh I'll do that, thanks. What about my idea to only target one version first and after that's accepted port to the other one?

pycbouh commented:
If you make it for master first, that's a preferred way to contribute any time.

dploeger commented:
Gotcha.

willnationsdev commented:
@dploeger Rebased this branch too, if it proves to be relevant at all.

dploeger commented:
This PR will be stalled until #48201 is merged, so I don't need to do things twice.

bfloch commented:
With 3.5 being almost out of the door, I hope the maintainers can comment if there is a chance for this to make it into a 3.x branch.

It would be such a welcome addition in order to round off a stable Godot version while 4.0 is being ironed out - even for existing projects, I'd pick it up immediately since without it I tend to have to write editor extensions just to be able to easily handle custom resources.

Thanks for the hard work.

dploeger commented:
I'm still waiting for a review on the 4.0.0 PR. I will see, whether it makes sense to backport it then.

quantumcode-martin commented:
That is my most wanted feature for 3.5.
I think Resources are way under considered and underrated, and this would bring custom resources to a new level.
It will definitely boost the usage and creation of custom resources.
I'll make a video explaining it and why it's so powerful as soon as it's available.

Will it be ready for 3.5 ?
Can we do something to speedup the integration and make it available faster ?

h0lley commented:
> Will it be ready for 3.5 ?

see the comment right above yours.
3.5 seems kind of unlikely at this rate, but there's still a chance I guess.

> Can we do something to speedup the integration and make it available faster ?

https://github.com/godotengine/godot/pull/48201 this needs to get finalized and merged first.

andersmmg commented:
Are there any updates on whether this will start to be backported again now that most of the 4.0 version is merged?

Atlinx commented:
Hey all, if anyone's interested I have this PR rebased and cherry-picked onto my personal Godot 3.x branch, and I'm automatically building my custom Godot version using Github Actions here: https://github.com/Atlinx/Godot3CustomBuilds. Feel free to download a build if you'd like.

akien-mga commented:
> Hey all, if anyone's interested I have this PR rebased and cherry-picked onto my personal Godot 3.x branch, and I'm automatically building my custom Godot version using Github Actions here: [Atlinx/Godot3CustomBuilds](https://github.com/Atlinx/Godot3CustomBuilds). These builds have both GDScript and C# support -- feel free to download a build if you'd like.

Linking to it directly for people who would want to backport it to their own forks: https://github.com/Atlinx/godot/commit/02d1f70ee5ec1b7bb08bc6dddac7fe39b4727e1e

---

At the current stage, I don't expect this PR to be merged in `3.x`, as it's a very significant rework that had a lot of implications in the `master` branch for 4.0, went through multiple iterations and several months of time and many bugfix PRs. We don't have the resources to do the same effort on the `3.x` branch.

So I'll close this not to give false expectations on the likelihood of this feature being included in 3.6 or later. Anyone who absolutely needs the feature can try the rebased version linked by Atlinx in their custom builds.
